File: lec2prac/Chamber of Secrets.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INF 1000000000

typedef struct {
    int r, c, d;
} State;

/* Deque implementation */
typedef struct {
    State *data;
    int head, tail, size;
} Deque;

void init_deque(Deque *dq, int cap) {
    dq->data = (State *)malloc(sizeof(State) * cap);
    dq->head = dq->tail = 0;
    dq->size = cap;
}

int empty(Deque *dq) {
    return dq->head == dq->tail;
}

void push_front(Deque *dq, State x) {
    dq->head = (dq->head - 1 + dq->size) % dq->size;
    dq->data[dq->head] = x;
}

void push_back(Deque *dq, State x) {
    dq->data[dq->tail] = x;
    dq->tail = (dq->tail + 1) % dq->size;
}

State pop_front(Deque *dq) {
    State x = dq->data[dq->head];
    dq->head = (dq->head + 1) % dq->size;
    return x;
}

int in_bounds(int r, int c, int n, int m) {
    return r >= 0 && r < n && c >= 0 && c < m;
}

int solution(char **grid, int n, int m) {
    // directions: up, right, down, left
    int dirs[4][2] = {{-1,0},{0,1},{1,0},{0,-1}};

    // dist[r][c][dir]
    int ***dist = (int ***)malloc(n * sizeof(int **));
    for (int i = 0; i < n; i++) {
        dist[i] = (int **)malloc(m * sizeof(int *));
        for (int j = 0; j < m; j++) {
            dist[i][j] = (int *)malloc(4 * sizeof(int));
            for (int d = 0; d < 4; d++)
                dist[i][j][d] = INF;
        }
    }

    Deque dq;
    init_deque(&dq, n * m * 4);

    // start at bottom-right going left (dir = 3)
    dist[n-1][m-1][3] = 0;
    push_back(&dq, (State){n-1, m-1, 3});

    while (!empty(&dq)) {
        State cur = pop_front(&dq);
        int r = cur.r, c = cur.c, d = cur.d;
        int cost = dist[r][c][d];

        // move straight
        int nr = r + dirs[d][0];
        int nc = c + dirs[d][1];
        if (in_bounds(nr, nc, n, m) && cost < dist[nr][nc][d]) {
            dist[nr][nc][d] = cost;
            push_front(&dq, (State){nr, nc, d});
        }

        // reflect at column
        if (grid[r][c] == '#') {
            for (int nd = 0; nd < 4; nd++) {
                if (nd == d) continue;
                int new_cost = cost + 1;
                if (new_cost < dist[r][c][nd]) {
                    dist[r][c][nd] = new_cost;
                    push_back(&dq, (State){r, c, nd});
                }
            }
        }
    }

    // answer: reach row 0 going left
    int ans = INF;
    for (int j = 0; j < m; j++) {
        if (dist[0][j][3] < ans)
            ans = dist[0][j][3];
    }

    return (ans == INF ? -1 : ans);
}

int main() {
    int n, m;
    scanf("%d %d", &n, &m);

    char **grid = (char **)malloc(n * sizeof(char *));
    for (int i = 0; i < n; i++) {
        grid[i] = (char *)malloc(m + 1);
        scanf("%s", grid[i]);
    }

    printf("%d\n", solution(grid, n, m));
    return 0;
}

File: cs33/lec2prac/Chamber of Secrets.py
"""
We model the movement of the basilisk's gaze as a shortest-path problem on a graph.
Each state is defined as (r,c,d) where
- (r,c) is a cell in the grid, and
- d is the direction the gaze is currently travelling.

We maintain a 3D array dist[r][c][d], which stores
the minimum number of columns turned magic needed
for the gaze to reach cell (r, c) while moving in
direction d.

The algorithm starts from the basilisk's position
at the bottom-right cell, initially facing left.
From any state,
- Move straight in the current direction with cost 0.
- Reflect at a column by changing direction, cost 1

Since all edge costs are either 0 or 1, we use a
BFS with a deque:
- transitions with cost 0 are pushed to front
- transitions with cost 1 are pushed to back

After the BFS finishes, we take the minimum cost
among all states that reach row 0 while moving left,
which corresponds to the gaze reaching the chamber 
entrance.

If no such state is reachable, the answer is -1.
"""

"""
Let the grid have dimensions n x m.
- Number of states: 4nm, each cell with 4 directions
- Each state is processed at most once with its final distance.
Time Complexity  = O(nm)
Space Complexity = O(nm)
for the distance array and deque.
"""
"""
 
reworded problem statement:
what is the MINIMUM number of columns we can make
magic so that the gaze can reach a column in row 0?
 
idea: if walang column on the first row to go left, 
its impossible to secure
 
problem details:
. if empty
# if regular column
"""
import sys
from collections import deque
 
def solution(grid: list[list[str]]) -> int:
    n, m = len(grid), len(grid[0])
    def in_bounds(i: int, j: int) -> bool:
        return 0 <= i < n and 0 <= j < m
    
    # up right down left
    dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]
 
    # dist[r][c][dir] = minimum cost to reach (r, c) going in direction dir
    dist = [[[float("inf")] * 4 for _ in range(m)] for _ in range(n)]
 
    # step 1: start from (n - 1, m - 1) going left
    dist[n - 1][m - 1][3] = 0
    dq = deque()
    dq.append((n - 1, m - 1, 3))
 
    while dq:
        r, c, d = dq.popleft()
        cost = dist[r][c][d]
 
        # move straight
        dr, dc = dirs[d]
        nr, nc = r + dr, c + dc
        if in_bounds(nr, nc) and cost < dist[nr][nc][d]:
                dist[nr][nc][d] = cost
                dq.appendleft((nr, nc, d)) # 0 cost, appendleft
 
        # reflect at any cell
        if grid[r][c] == "#":
            for nd in range(4):
                if nd == d:
                    continue
                # turning at a column or empty cell costs + 1
                new_cost = cost + 1
                if new_cost < dist[r][c][nd]:
                    dist[r][c][nd] = new_cost
                    dq.append((r, c, nd))
         
    # find minimum cost to reach any column on row 0
    ans = float("inf")
    for j in range(m):
        ans = min(ans, dist[0][j][3])
 
    return -1 if ans == float("inf") else ans
 
def main() -> None:
    input = sys.stdin.readline
 
    n, m = map(int, input().split())
 
    grid = [list(input().strip()) for _ in range(n)]
 
    print(solution(grid))    
 
if __name__ == "__main__":
    main()